src/zql/query/EntityQuery.ts

```diff
@@ -0,0 +1,121 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import {Misuse} from '../error/Misuse.js';
+import {EntitySchema} from '../schema/EntitySchema.js';
+import {IEntityQuery, Selectable, SelectedFields} from './IEntityQuery.js';
+import {IStatement, Statement} from './Statement.js';
+import {AST, Operator, Primitive} from '../ast/ZqlAst.js';
+import {Context} from '../context/context.js';
+
+let aliasCount = 0;
+
+export class EntityQuery<S extends EntitySchema, TReturn = []>
+  implements IEntityQuery<S, TReturn>
+{
+  readonly #ast: AST;
+  readonly #name: string;
+  readonly #context: Context;
+
+  constructor(context: Context, tableName: string, ast?: AST) {
+    this.#ast = ast ?? {
+      table: tableName,
+      alias: aliasCount++,
+    };
+    this.#name = tableName;
+    this.#context = context;
+  }
+
+  select<Fields extends Selectable<S>[]>(...x: Fields) {
+    if (this.#ast.select === 'count') {
+      throw new Misuse(
+        'A query can either return fields or a count, not both.',
+      );
+    }
+
+    return new EntityQuery<S, SelectedFields<S['fields'], Fields>[]>(
+      this.#context,
+      this.#name,
+      {
+        ...this.#ast,
+        select: [...new Set([...(this.#ast.select || []), ...x])] as any,
+      },
+    );
+  }
+
+  where<K extends keyof S['fields']>(
```

Comment:
I had been thinking you would have to select a field in order to filter on it in where. Of course that is not how SQL works, and I think what you have here makes more sense. It will complicate slightly the logic for determining which columns to sync to the client.

Reply:
We will also need logic for determining the 'required columns' for a query on the client, so that we can filter the source to just entities that have the 'required columns'.  I'm imagining this would be the very first step in the pipeline.